python/ray/data/_internal/execution/streaming_executor_state.py

```diff
@@ -645,4 +648,14 @@ def _execution_allowed(
         object_store_memory=downstream_usage.object_store_memory
     ).satisfies_limit(downstream_limit)
 
+    # If completing a task decreases the overall object store memory usage, allow it
+    # even if we're over the global limit.
+    if (
+        DataContext.get_current().use_runtime_metrics_scheduling
+        and global_ok_sans_memory
+        and op.metrics.average_bytes_change_per_task is not None
+        and op.metrics.average_bytes_change_per_task <= 0
+    ):
+        return True
+
     return global_ok_sans_memory and downstream_memory_ok
```

Comment:
```suggestion
    if DataContext.get_current().use_runtime_metrics_scheduling:
        return (
            global_ok_sans_memory
            and op.metrics.average_bytes_change_per_task is not None
            and op.metrics.average_bytes_change_per_task <= 0
        )
    else:
        return global_ok_sans_memory and downstream_memory_ok
```

Reply:
I tested this out with `test_large_e2e_backpressure`, and doing an `if` / `else` might not work. We're over the memory limit and `average_bytes_change_per_task` is initially `None`, so `consume` tasks aren't launched; as a result, we keep pulling data from the `produce` tasks until all of the `produce` tasks are complete.

Left this code as the original change for now. Hopefully that should minimize the risk, while still providing some benefit. In the long term, I think we need to devise a strategy for when metrics are `None`.